### DISCUSSION TITLE
Can't send email: ERROR 504

### DISCUSSION THREAD

**Discussion by romu70** [dani-garcia/vaultwarden#5019] [Q&A] [answered]
I see my VW instance can't send emails.
I've got to the admin page to make a small testing, and I get this error. I don't know if this means something for someone here.

![Screenshot 2024-10-01 at 14 12 04](https://github.com/user-attachments/assets/648c8b9a-d1a4-45f7-83ae-2814124c8458)

I've tried with gomtp, and it does send emails with the setting I use for VW.
Any help would be great.

**Comment by BlackDex** [comment]
enable `SMTP_DEBUG=true` and check the logs for details.

**Reply by romu70** [reply]
Thanks.

2 points:

1. It runs in a Docker container with docker-compose, is this a setting accessible from the UI or I need to modify the docker-compose command?
2. 504 seems to be a timeout, I've been warned to increase my KDF iterations number which was 100000. Now the minimum is 600000. But my server is an old Atom processor, and I wonder of this could be the reason of the timeout.

**Reply by BlackDex** [reply]
504 is not a timeout in the sense of an SMTP server, totally different meanings.
It is an environment variable, same as all settings, See:
https://github.com/dani-garcia/vaultwarden/blob/040e2a7bb0f2cc5012d46ca99283cf21fa06ed1a/.env.template#L540-L543

**Reply by romu70** [reply]
I tweaked the docker-compose file to enable the SMTP debug. The log says:

> [WARNING] SMTP Debugging is enabled (SMTP_DEBUG=true). Sensitive information could be disclosed via logs!

So I guess it's on. And when the popup appears, there is nothing in the log.

**Reply by BlackDex** [reply]
If the smtp debug is set to true, it should show the whole smtp dialog with the smtp server.

So you should see something, if not then smtp isn't triggered

**Reply by romu70** [reply]
Is there something wrong in there?

![Screenshot 2024-10-02 at 09 09 32](https://github.com/user-attachments/assets/06127547-1d56-4bb8-90fa-886d5ff73fd7)

And as I wrote, the log says SMTP_DEBUG is enabled, but nothing is logged when I click on the "Send test email" button. There is only the popup which appears after around 30s.

**Reply by BlackDex** [reply]
Well, then in that case there is a timeout indeed, and it looks like the call never reaches Vaultwarden. So i suggest to look at your reverse proxy.

**Reply by BlackDex** [reply]
Also, i notice that you have `starttls` and port 465, that is most of the time not the right combination.
465 is TLS/SSL, while 587 is `starttls`

**Reply by romu70** [reply]
Thanks.
465 is not the problem, it works with gomtp, and they are the settings given by my provider.

I'll look at the reverse proxy, but not sure what should I look exactly. Any idea?

My server is a Synology NAS, and the reverse proxy is the one provided by Synology as well.

**Reply by romu70** [reply]
From my nas, with gomtp, I'm able to send emails with these settings.

**Reply by romu70** [reply]
From within the VW container, using gomtp, I'm able to send emails with these settings. So I would say this is not a reverse proxy issue, but a problem with VW.

**Reply by BlackDex** [reply]
That is not comparable. If Vaultwarden does not receive the command to send the mail, how can it start sending the mail.
If you do not see any logs, then Vaultwarden does not receive the request to do so.

Sure, from within the container it will probably work, but if the reverse proxy is not passing on the request to Vaultwarden, then the problem is at your reverse proxy.

Also, set `LOG_LEVEL=debug` then, and you MUST see logs appear, if not, then it is **not** a problem of Vaultwarden.

**Reply by romu70** [reply]
I see this in the log:

> [2024-10-02 08:51:55.691][request][INFO] POST /admin/test/smtp

**Reply by BlackDex** [reply]
And after that nothing?
Not even with `LOG_LEVEL=debug` ?

Please provide the `/admin/diagnostics` support string so that we have a bit more details.

**Reply by romu70** [reply]
With the Firefox debugger, I see the smtp message which gets a 504 error. In the response, instead of a Json, it's the HTML of the "Page not found" Synology page. So I think you're right, it's probably the reverse proxy. But I don't know how to fix this at all.

Here is the diagnostics data:

> ### Your environment (Generated via diagnostics page)
* Vaultwarden version: v1.32.0
* Web-vault version: v2024.6.2b
* OS/Arch: linux/x86_64
* Running within a container: true (Base: Debian)
* Environment settings overridden: false
* Uses a reverse proxy: true
* IP Header check: true (X-Real-IP)
* Internet access: true
* Internet access via a proxy: false
* DNS Check: true
* Browser/Server Time Check: true
* Server/NTP Time Check: true
* Domain Configuration Check: true
* HTTPS Check: true
* Database type: PostgreSQL
* Database version: PostgreSQL 15.2 on x86_64-pc-linux-musl, compiled by gcc (Alpine 12.2.1_git20220924-r4) 12.2.1 20220924, 64-bit
* Clients used: 
* Reverse proxy and version: 
* Other relevant information: 

### Config (Generated via diagnostics page)
<details><summary>Show Running Config</summary>

**Environment settings which are overridden:** 


```json
{
  "_duo_akey": null,
  "_enable_duo": true,
  "_enable_email_2fa": true,
  "_enable_smtp": true,
  "_enable_yubico": true,
  "_icon_service_csp": "",
  "_icon_service_url": "",
  "_ip_header_enabled": true,
  "_max_note_size": 10000,
  "_smtp_img_src": "cid:",
  "admin_ratelimit_max_burst": 3,
  "admin_ratelimit_seconds": 300,
  "admin_session_lifetime": 20,
  "admin_token": "***",
  "allowed_iframe_ancestors": "",
  "attachments_folder": "data/attachments",
  "auth_request_purge_schedule": "30 * * * * *",
  "authenticator_disable_time_drift": false,
  "data_folder": "data",
  "database_conn_init": "",
  "database_max_conns": 10,
  "database_timeout": 30,
  "database_url": "**********://*************************************************************",
  "db_connection_retries": 0,
  "disable_2fa_remember": false,
  "disable_admin_token": false,
  "disable_icon_download": false,
  "domain": "*****://****************",
  "domain_origin": "*****://****************",
  "domain_path": "",
  "domain_set": true,
  "duo_context_purge_schedule": "30 * * * * *",
  "duo_host": null,
  "duo_ikey": null,
  "duo_skey": null,
  "duo_use_iframe": false,
  "email_2fa_auto_fallback": false,
  "email_2fa_enforce_on_verified_invite": false,
  "email_attempts_limit": 3,
  "email_change_allowed": true,
  "email_expiration_time": 600,
  "email_token_size": 6,
  "emergency_access_allowed": true,
  "emergency_notification_reminder_schedule": "0 3 * * * *",
  "emergency_request_timeout_schedule": "0 7 * * * *",
  "enable_db_wal": true,
  "enable_websocket": true,
  "event_cleanup_schedule": "0 10 0 * * *",
  "events_days_retain": null,
  "experimental_client_feature_flags": "fido2-vault-credentials",
  "extended_logging": true,
  "helo_name": null,
  "hibp_api_key": null,
  "http_request_block_non_global_ips": true,
  "http_request_block_regex": null,
  "icon_blacklist_non_global_ips": true,
  "icon_blacklist_regex": null,
  "icon_cache_folder": "data/icon_cache",
  "icon_cache_negttl": 259200,
  "icon_cache_ttl": 2592000,
  "icon_download_timeout": 10,
  "icon_redirect_code": 302,
  "icon_service": "internal",
  "incomplete_2fa_schedule": "30 * * * * *",
  "incomplete_2fa_time_limit": 3,
  "increase_note_size_limit": false,
  "invitation_expiration_hours": 120,
  "invitation_org_name": "Vaultwarden",
  "invitations_allowed": true,
  "ip_header": "X-Real-IP",
  "job_poll_interval_ms": 30000,
  "log_file": "/data/bitwarden.log",
  "log_level": "debug",
  "log_timestamp_format": "%Y-%m-%d %H:%M:%S.%3f",
  "login_ratelimit_max_burst": 10,
  "login_ratelimit_seconds": 60,
  "org_attachment_limit": null,
  "org_creation_users": "",
  "org_events_enabled": false,
  "org_groups_enabled": false,
  "password_hints_allowed": true,
  "password_iterations": 600000,
  "push_enabled": false,
  "push_identity_uri": "https://identity.bitwarden.com",
  "push_installation_id": "***",
  "push_installation_key": "***",
  "push_relay_uri": "https://push.bitwarden.com",
  "reload_templates": false,
  "require_device_email": false,
  "rsa_key_filename": "data/rsa_key",
  "send_purge_schedule": "0 5 * * * *",
  "sendmail_command": null,
  "sends_allowed": true,
  "sends_folder": "data/sends",
  "show_password_hint": false,
  "signups_allowed": true,
  "signups_domains_whitelist": "",
  "signups_verify": false,
  "signups_verify_resend_limit": 6,
  "signups_verify_resend_time": 3600,
  "smtp_accept_invalid_certs": false,
  "smtp_accept_invalid_hostnames": false,
  "smtp_auth_mechanism": null,
  "smtp_debug": true,
  "smtp_embed_images": true,
  "smtp_explicit_tls": null,
  "smtp_from": "********************",
  "smtp_from_name": "Vaultwarden",
  "smtp_host": "*******************",
  "smtp_password": "***",
  "smtp_port": 465,
  "smtp_security": "starttls",
  "smtp_ssl": null,
  "smtp_timeout": 15,
  "smtp_username": "********************",
  "templates_folder": "data/templates",
  "tmp_folder": "data/tmp",
  "trash_auto_delete_days": null,
  "trash_purge_schedule": "0 5 0 * * *",
  "use_sendmail": false,
  "use_syslog": false,
  "user_attachment_limit": null,
  "user_send_limit": null,
  "web_vault_enabled": true,
  "web_vault_folder": "web-vault/",
  "yubico_client_id": null,
  "yubico_secret_key": null,
  "yubico_server": null
}
```
</details>

**Reply by BlackDex** [reply]
Config looks fine as far as i can tell.
I'm also not an expert on Synology hardware maybe someone else here, else i would suggest to check a Synology forum somewhere.

**Reply by romu70** [reply]
Could you please describe which requests are blocked in there? Protocols, addresses, ports, etc. This would be really helpful to diagnose my setup.

**Reply by BlackDex** [reply]
Well i don't know. The strange thing is that you can access the interface. So no clue as to why it doesn't go through.
It's just exactly the same as all other requests.

Just check your reverse proxy logs, try to use the developer console of the browser or something like that.

**Reply by romu70** [reply, marked as answer]
Finale made il work 🕺.
I switched from ```starttls``` to ```force_tls``` in the email settings.

Thanks a lot for your help @BlackDex.

**Reply by Gerardv514** [reply]
> Also, i notice that you have `starttls` and port 465, that is most of the time not the right combination. 465 is TLS/SSL, while 587 is `starttls`

I use starttls and i have no problem; also using port 587. As blackdex mentioned, according to your initial screenshot you have starttls with port 465; this is wrong and should be 587 with this combination. 

If you want port 465 than as you have resolved; force tls should be used.

**Reply by romu70** [reply]
Thanks. My provider forces its users to use 465.

**Reply by BlackDex** [reply]
But then the settings with gomtp and Vaultwarden were not the same i guess. Or gomtp does some other fancy stuff.

**Reply by romu70** [reply]
gompt settings:

```
port: 465
ssl: true
tls: true
```
